docs: add overview of zkVM fault proof (#20)
kroma-network/kroma-specs: specs/experimental/zk-fault-dipute-game/overview.md

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

> [g-withdrawal]: ../../glossary.md#withdrawal

Reviewer: ```suggestion
[g-withdrawals]: ../../glossary.md#withdrawals
```

## Overview

The ZK Fault Dispute Game (ZKFDG) is a new type of dispute game compatible with the OP Stack. In ZKFDG, the first

> disagreeing output root is found by the process called "dissection". Then, both derivation of L2 block from L1

Reviewer: How about replacing `output root` with `root claim` in this document?

Author: Ok. I'll change it. But output root is also used in dispute game. They use both terms.

> disagreeing output root is found by the process called "dissection". Then, both derivation of L2 block from L1
> corresponding to that output root and executing the block's transactions is executed in zkVM. The validity of the

Reviewer: ```suggestion
disagreeing output root between defender and challenger is found by the process called "dissection". Then, both the derivation of L2 block
corresponding to that output root and the execution of the block's transactions are executed in zkVM. The validity of the
```

execution is guaranteed by the ZK proof, which can be verified on-chain.

The main difference from [Kroma's previous ZK Fault Proof System](../../fault-proof/challenge.md) is that ZKFDG uses a

> "zkVM" instead of a "zkEVM" to handle all processes. By using zkVM for fault proof, it is enabled to verify all

Reviewer: ```suggestion
"zkVM" instead of a "zkEVM" to handle the proving fault process. By using zkVM for fault proof, it is possible to verify the entire
```

processes from derivation to execution without any additional developments of ZK circuit.

> ZKFDG implements the OP Stack's `IDisputeGame`, implying that it is fully compatible with the OP Stack's dispute game.

Reviewer: ```suggestion
ZKFDG implements the OP Stack's `IDisputeGame`, implying that it is fully compatible as one of the OP Stack's dispute game types.
```

This also implies that it can be applied to the OP Stack's multi-proof system in the future.

## ZK Fault Dispute Game Creation

The ZK Fault Dispute Game is created by a validator selected by the

> [`ValidatorManager`](../../protocol/validator-v2/validator-manager.md) at intervals defined by the

Reviewer: ```suggestion
[`ValidatorManager`](../../protocol/validator-v2/validator-manager.md) at intervals defined as the
```

`SUBMISSION_INTERVAL` configured in the [`L2OutputOracle`](../../protocol/validation.md#l2-output-oracle-smart-contract)
contract. The validator can create a ZKFDG-type dispute game through the `DisputeGameFactory` with an extra data. The
extra data is composed as follows:

```text
extra_data = current_l2_block_num ++ prev_game_proxy

current_l2_block_num = uint256
prev_game_proxy      = bytes32
```

When the ZKFDG is created by the validator, it is verified that the validator is eligible to submit a claim and the

> `prev_game_proxy` is valid ZKFDG contract. The `prev_game_proxy` is used as the starting output root for the dissection
> process to find the first disagreeing block. If all validations pass, the address of this ZKFDG contract is stored with

Reviewer: ```suggestion
`prev_game_proxy` is the right previous ZKFDG contract. The root claim stored in `prev_game_proxy` is used as the starting output root for the dissection
process to find the first disagreeing block. If all validations are passed, the address of this ZKFDG contract is stored with
```

Reviewer: Don't we need the content about current_l2_block_num validation?

Author: I'll add it.

current L2 block number on the L2OutputOracle contract. This will be used for the creation of the next ZKFDG.

## Challenge Creation

> As with Kroma's previous Fault Proof System, a challenge can be initiated by a validator who disagrees the submitted

Reviewer: How about linking to challenge creation section in previous challenge document?

> claim. The challenge process begins by submitting the initial segments by challenger.

Reviewer: ```suggestion
claim. The challenge process begins by submitting the intermediate segments between starting root claim and disputed root claim by challenger.
```

```solidity
/**
     * @notice Creates a challenge against an invalid claim.
     *
     * @param _l1BlockHash   The block hash of L1 at the time the output L2 block was created.
     * @param _l1BlockNumber The block number of L1 with the specified L1 block hash.

>      * @param _segments      Array of the segment. A segment is the first output root of a specific range.

Reviewer: ```suggestion
     * @param _segments      Array of the segments. A segment is the first output root of a specific range.
```

     */
    function createChallenge(
        bytes32 _l1BlockHash,
        uint256 _l1BlockNumber,
        bytes32[] calldata _segments
    ) external;
```

## Dissection

> The dissection process is carried out in the same manner as in Kroma's previous Fault Proof System. Through interactions

Reviewer: How about linking to bisection section in previous challenge document?

between the challenger and the defender (the game creator), the first disagreeing output root can be specified.

```solidity
    /**

>      * @notice Selects an invalid section and submit segments of that section.

Reviewer: ```suggestion
     * @notice Selects an invalid section and submits intermediate segments of that section.
```

     *
     * @param _challenger  Address of the challenger.
     * @param _pos         Position of the last valid segment.

>      * @param _segments    Array of the segment. A segment is the first output root of a specific range.

Reviewer: ```suggestion
     * @param _segments    Array of the segments. A segment is the first output root of a specific range.
```

     */
    function dissect(
        address _challenger,
        uint256 _pos,
        bytes32[] calldata _segments
    ) external;
```

> ## Proving Fault using ZKVM

Reviewer: Maybe here some contents will be added, as mentioned in [this comment](https://github.com/kroma-network/kroma-specs/pull/19#pullrequestreview-2268055565), right?

Author: Yes you're right

Author: done in #21

## Resolution of ZK Fault Dispute Game

For the ZKFDG to be resolved, the `FINALIZATION_PERIOD` must pass after it is created. After the `FINALIZATION_PERIOD`

> has passed, the status of dispute game can be determined based on the presence of the claim root in the dispute game. If

Reviewer: How about unifying `claim root` to `root claim` in this section?

a non-zero value is stored for the claim root, the dispute game's status will be resolved as `DEFENDER_WINS`. If the

> claim root value is stored with zero, it indicates that the challenger succeeded in zk proving for fault, and the
> dispute game will be resolved as `CHALLENGER_WINS`. The resolved ZKFDG is used for [withdrawal][g-withdrawal] to L1.

Reviewer: ```suggestion
claim root value is stored with zero, it indicates that the challenger succeeded in zk proving for fault, thus the
dispute game will be resolved as `CHALLENGER_WINS`. Only ZKFDG resolved as `DEFENDER_WINS` can be used for proving and finalizing [withdrawals][g-withdrawals] to L1.
```

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

[g-bisection]: ../../../specs/fault-proof/challenge.md#bisection

> [g-challenge-creation]: ../../../specs/fault-proof/challenge.md#challenge-creation

Reviewer: These are not glossary. Please rename these to `[prev-bisection]`, `[prev-challenge-creation]`. Also move these below each corresponding usage section.

> [g-withdrawals]: ../../glossary.md#withdrawals

Reviewer: Could you move up these parts right below the heading with below comment?
```<!-- All glossary references in this file. -->```

## Overview

The ZK Fault Dispute Game (ZKFDG) is a new type of dispute game compatible with the OP Stack. In ZKFDG, the first
disagreeing output root between defender and challenger is found by the process called "dissection". Then, both the
derivation of L2 block corresponding to that output root and the execution of the block's transactions are carried out
in zkVM. The validity of the execution is guaranteed by the ZK proof, which can be verified on-chain.

The main difference from [Kroma's previous ZK Fault Proof System](../../fault-proof/challenge.md) is that ZKFDG uses a
"zkVM" instead of a "zkEVM" to handle the proving fault process. By using zkVM for fault proof, it is possible to verify
the entire processes from derivation to execution without any additional developments of ZK circuit.

ZKFDG implements the OP Stack's `IDisputeGame`, implying that it is fully compatible as one of the OP Stack's dispute
game type. This also implies that it can be applied to the OP Stack's multi-proof system in the future.

## ZK Fault Dispute Game Creation

> The ZK Fault Dispute Game is created by a validator selected as the
> [`ValidatorManager`](../../protocol/validator-v2/validator-manager.md) at intervals defined by the

Reviewer: ```suggestion
The ZK Fault Dispute Game is created by a validator selected by the
[`ValidatorManager`](../../protocol/validator-v2/validator-manager.md) at intervals defined as the
```

- the validator should be eligible to submit a claim
- The `current_l2_block_num` should be the next block number for submission of root claim (output root)
- The `prev_game_proxy` should be the right previous ZKFDG contract

If all validations are passed, a new ZKFDG contract is created. The address of this ZKFDG contract is stored with
current L2 block number on the L2OutputOracle contract, which can be used for the creation of the next ZKFDG. The root
claim stored in `prev_game_proxy` is used as the starting output root for the dissection process to find the first
disagreeing block.

## Challenge Creation

As with Kroma's [previous Fault Proof System][g-challenge-creation], a challenge can be initiated by a validator who
disagrees the submitted claim. The challenge process begins by submitting the intermediate segments between starting
output root and disputed output root by challenger.

```solidity
/**
     * @notice Creates a challenge against an invalid claim.
     *
     * @param _l1BlockHash   The block hash of L1 at the time the output L2 block was created.
     * @param _l1BlockNumber The block number of L1 with the specified L1 block hash.

>      * @param _segments      Array of the segments. A segment is the first output of a specific range.

Reviewer: ```suggestion
     * @param _segments      Array of the segments. A segment is the first output root of a specific range.
```
